### PR TITLE
Simple Android client

### PR DESCRIPTION
**Pull request by dmdashenkov**
Introducing TodoList Android client.

The current implementation of the client has 2 screens:
 1. One displaying the `MyListView` projection state.
 2. Other for creating new `Task`s.

The application uses subscription mechanism to receive updates of the `MyListView` projection state.

Also, to make it possible to share common code base between the `client` module and the Android application project, now we publish the `client` and `model` modules to Maven.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: android-app/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# TodoList Android Client
```
Comment (armiol) [inline android-app/README.md]: Let's call it `Todo List` . Here and below.

File: android-app/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+To run the application, open the `android-app` module in the Android Studio and launch the app with
+the default application run configuration.
+
+Alternatively, build the app as described above, go to 
+[`./app/build/outputs/apk/debug/app-debug.apk`](./app/build/outputs/apk/debug/app-debug.apk), copy 
+the `.apk` file to an Android device and install and start the application.
+
+### Connection to the Server
+
+By default, the application connects to the localhost server started on the port `50051`.
```
Comment (armiol) [inline android-app/README.md]: `localhost` should go in monospaced font.

File: android-app/app/build.gradle [hunk 3]
```diff
@@ -9,20 +9,54 @@ android {
...
     buildTypes {
         release {
             minifyEnabled false
             proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
         }
     }
+    compileOptions {
+        /*
+            As TodoList uses Java 8.
+            This is not obligatory for any Spine project that uses Java 7.
```
Comment (armiol) [inline android-app/app/build.gradle]: This sentence does not make much sense.

File: android-app/app/build.gradle [hunk 4]
```diff
@@ -9,20 +9,54 @@ android {
...
+        */
+        targetCompatibility 1.8
+        sourceCompatibility 1.8
+    }
+
+    packagingOptions {
+        /*
+            Spine auto generated .property files.
```
Comment (armiol) [inline android-app/app/build.gradle]: `auto-generated` will look better.

File: android-app/app/build.gradle [hunk 5]
```diff
@@ -9,20 +9,54 @@ android {
...
+            no simple way to avoid those duplicates. Though, their presence does not much harm to
+            the application, except for the larger APK file sizes.
+        */
+        merge 'known_types.properties'
+        merge 'validation_rules.properties'
+    }
+
+    lintOptions {
+        /*
+            gRPC uses some `javax` packages that are not included into the Android SDK. Though,
```
Comment (armiol) [inline android-app/app/build.gradle]: I still don't understand why `though`. 

Usually it's like

```
A isn't very good. B is acceptable, though.
```
Or
```
Though A isn't very good, B is still acceptable.
```
And, most importantly, you don't make this comment explaining the actual line. I.e. `abortOnError false`.

File: android-app/app/src/main/java/io/spine/examples/todolist/connection/Clients.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+     * Retrieves an instance of {@link SubscribingTodoClient} connecting to the gRPC server at
+     * {@link #HOST}{@code :}{@link #PORT}.
+     *
+     * <p>Do not call
+     * {@link io.spine.examples.todolist.client.TodoClient#shutdown() shutdown()} on the instances
+     * returned by this method.
+     *
+     * @return an instance of subscribing TodoList client
+     */
+    public static SubscribingTodoClient instance() {
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/connection/Clients.java]: If that's a factory of clients (plural), I'd avoid having a method names `instance`. With this naming it's impossible to have other clients.

File: android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/BaseActivity.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+ *
+ * <p>Any {@code Activity} in the app has an associated {@link BaseViewModel ViewModel} instance.
+ * The view model performs all the logic of connecting the view (UI) and the model (gRPC client).
+ *
+ * <p>An {@code Activity} calls the {@code ViewModel} methods to retrieve the data to display,
+ * passes the data from the UI, subscribes on the data updates, etc.
+ *
+ * @param <VM> the type of the {@code BaseViewModel} associated with this {@code Activity}
+ */
+public abstract class BaseActivity<VM extends BaseViewModel> extends AppCompatActivity {
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/BaseActivity.java]: I don't understand, what is so `Base` about those. Can't we just have `AbstractActivity` and `AbstractViewModel`?

File: android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/BaseViewModel.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+import static io.grpc.Status.Code.UNAVAILABLE;
+
+/**
+ * An implementation base for all the {@code ViewModel}s in the app.
+ *
+ * <p>The {@code ViewModel} classes hold data, perform network connections, etc. on the behalf of
+ * the associated view elements (such as {@code Activity}).
+ */
+@SuppressWarnings("AbstractClassWithoutAbstractMethods")
+    // API detail.
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/BaseViewModel.java]: Let's move that to the previous line AND rephrase.

File: android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/BaseViewModel.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+     * The TodoList gRPC client.
+     */
+    private final SubscribingTodoClient client = Clients.instance();
+
+    /**
+     * The {@link ExecutorService} performing the asynchronous operations, such as networking.
+     *
+     * @see #execute(Runnable) for the application
+     */
+    private final ExecutorService executor = Executors.newSingleThreadExecutor();
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/BaseViewModel.java]: This is a design smell. I'd not have anything as general-purpose as this executor.

Let's define a clear purpose for it and name/document accordingly.

File: android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/BaseViewModel.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+     * @param task the operation to execute
+     * @see ErrorCallback for the list of <i>supported</i> errors
+     */
+    protected void execute(Runnable task) {
+        checkNotNull(task);
+        executor.execute(() -> {
+            try {
+                task.run();
+            } catch (StatusRuntimeException e) {
+                Log.d(TAG, "execute: error task");
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/lifecycle/BaseViewModel.java]: Is this really the log level we should use for error logging.

File: android-app/app/src/main/java/io/spine/examples/todolist/mylist/MyListViewModel.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+    private static final String TAG = MyListViewModel.class.getSimpleName();
+
+    private final MutableLiveData<MyListView> myList = new MutableLiveData<>();
+
+    private Subscription myTasksSubscription;
+
+    // Required by the `ViewModelProviders` utility.
+    public MyListViewModel() {
+    }
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/mylist/MyListViewModel.java]: Let's have it on the previous line. We have this fixed in the code style previously — i.e. to make `{}` possible.  Please feel free to adjust this setting in this repository.

File: android-app/app/src/main/java/io/spine/examples/todolist/mylist/MyTaskListViewAdapter.java [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+ * The {@link RecyclerView.Adapter} implementation listening to the {@link MyListView} data updates
+ * and initiating the {@link RecyclerView} updates upon the data updates.
+ */
+final class MyTaskListViewAdapter
+        extends Adapter<MyTaskListViewAdapter.TaskViewHolder> implements Observer<MyListView> {
+
+    /**
+     * The map of the {@link LabelColor} enum to the RGB values of the color.
+     */
+    private static final Map<LabelColor, Integer> COLORS = ImmutableMap.of(
```
Comment (armiol) [inline android-app/app/src/main/java/io/spine/examples/todolist/mylist/MyTaskListViewAdapter.java]: Can't we reuse what's defined in the model? Let's discuss.

Comment (dmdashenkov) [reply android-app/app/src/main/java/io/spine/examples/todolist/mylist/MyTaskListViewAdapter.java]: Replaced with existing constants.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

**Pull request merged**